src/pretalx/submission/models/submission.py

```diff
@@ -288,39 +299,70 @@ def _set_state(self, new_state, force=False, person=None):
+        # qs_temp = TalkSlot.objects.filter(
+        #     submission=self,
+        #     schedule=self.event.wip_schedule,
+        # )
+        # logging.debug('TalkSlots old: {}'.format(
+        #     [i.slot_index for i in qs_temp]))
+        # logging.debug('self.state: {}'.format(self.state))
+        if self.state == SubmissionStates.ACCEPTED or \
+                self.state == SubmissionStates.CONFIRMED:
+            # logging.debug('state is ACCEPTED or CONFIRMED')
+            if slot_count_old is not None:
+                # logging.debug(
+                #     'slot_count_old: {}'.format(slot_count_old))
+                # logging.debug(
+                #     'self.slot_count {}'.format(self.slot_count))
+                range_to_delete = list(range(
+                    self.slot_count, slot_count_old))
+                # logging.debug('range_to_delete: {}'.format(range_to_delete))
+                qs_delete = TalkSlot.objects.filter(
+                    submission=self,
+                    schedule=self.event.wip_schedule,
+                    slot_index__in=range_to_delete,
+                )
+                # logging.debug('qs_delete: {}'.format(
+                #     [i.slot_index for i in qs_delete]))
+                qs_delete.delete()
+            # logging.debug('TalkSlot update_or_create')
+            for index in range(self.slot_count):
+                TalkSlot.objects.update_or_create(
+                    submission=self,
+                    schedule=self.event.wip_schedule,
+                    defaults={'is_visible': True},
+                    slot_index=index,
+                )
+        else:
+            TalkSlot.objects.filter(
+                submission=self, schedule=self.event.wip_schedule
+            ).delete()
+        # debug: print all now available Talks
+        # qs_temp = TalkSlot.objects.filter(
+        #     submission=self,
+        #     schedule=self.event.wip_schedule,
+        # )
+        # logging.debug('TalkSlots new: {}'.format(
+        #     [i.slot_index for i in qs_temp]))
 
-        TalkSlot.objects.filter(
-            submission=self, schedule=self.event.wip_schedule
-        ).delete()
+    def make_submitted(self, person=None, force=False, orga=False):
+        self._set_state(SubmissionStates.SUBMITTED, force, person=person)
 
     def confirm(self, person=None, force=False, orga=False):
         self._set_state(SubmissionStates.CONFIRMED, force, person=person)
         self.log_action('pretalx.submission.confirm', person=person, orga=orga)
-        self.slots.filter(schedule=self.event.wip_schedule).update(is_visible=True)
-        from pretalx.schedule.models import TalkSlot
 
-        TalkSlot.objects.update_or_create(
-            submission=self,
-            schedule=self.event.wip_schedule,
-            defaults={'is_visible': True},
-        )
+        self.slots.filter(schedule=self.event.wip_schedule).update(is_visible=True)
```

Comment:
Hm, we could consider doing this in the `update_talk_slots` method, semantically it seems appropriate.

Reply:
ok - i think iam missing something here...
if i fix the `update_or_create` call this should already do just what this line does?!